### PR TITLE
Public token fixes, API overhaul

### PR DESCRIPTION
**Pull request by AnIrishDuck**
- Move everything under V2
- Bring the api more in line with other public implementations
- Fix the public token issues
- Bump version to 0.2.0

@mguymon

### CODE REVIEW

File: lib/paseto/local.rb [hunk 1]
```diff
@@ -1,62 +1,69 @@
...
-    end
+      class Key
+        def self.generate
+          new(RbNaCl::Random.random_bytes(RbNaCl::AEAD::ChaCha20Poly1305IETF.key_bytes))
+        end
 
-    def self.generate_nonce
-      RbNaCl::Random.random_bytes(NONCE_BYTES)
-    end
+        def self.decode64(encoded_key)
```
Comment (mguymon) [inline lib/paseto/local.rb]: So I think this is a little bit of name collision for `Pasteo::V2::Local.decode64` vs `Pasteo.decode64`. I think it sides steps the possible confusion by keeping the names distinct.

File: lib/paseto/local.rb [hunk 2]
```diff
@@ -1,62 +1,69 @@
...
+          @aead = RbNaCl::AEAD::ChaCha20Poly1305IETF.new(key)
+        end
 
-    def initialize(private_key, footer = nil)
-      @aead = RbNaCl::AEAD::ChaCha20Poly1305IETF.new(private_key)
-      @footer = footer
-    end
+        def encode64
+          Paseto.encode64(@key)
+        end
```
Comment (mguymon) [inline lib/paseto/local.rb]: Does the impl needs to expose an aliased `encode64`?

File: lib/paseto/local.rb [hunk 3]
```diff
@@ -1,62 +1,69 @@
...
+
+        attr_reader :aead
+      end
 
-    def encrypt(message)
-      # Make a nonce: A single-use value never repeated under the same key
-      nonce = self.class.generate_nonce
+      def self.generate_nonce
```
Comment (mguymon) [inline lib/paseto/local.rb]: Stylistic comment: class methods should hang out together at the top of the class def

File: lib/paseto/local.rb [hunk 4]
```diff
@@ -1,62 +1,69 @@
...
 
-    def initialize(private_key, footer = nil)
-      @aead = RbNaCl::AEAD::ChaCha20Poly1305IETF.new(private_key)
-      @footer = footer
-    end
+        def encode64
+          Paseto.encode64(@key)
+        end
+
+        attr_reader :aead
```
Comment (mguymon) [inline lib/paseto/local.rb]: stylistic - accessor helpers should be at the top of the class unless private

File: lib/paseto/public.rb [hunk 5]
```diff
@@ -1,54 +1,84 @@
...
-      raise BadMessageError.new('Unable to process message') if decoded_message.nil? || signature.nil?
+        def encode64
+          Paseto.encode64(@key)
+        end
 
-      begin
-        @verify_key.verify(signature, decoded_message)
-      rescue RbNaCl::BadSignatureError
-        false
+        attr_reader :nacl
```
Comment (mguymon) [inline lib/paseto/public.rb]: stylistic - accessor helpers should be at the top of the class unless private

File: lib/paseto/public.rb [hunk 6]
```diff
@@ -1,54 +1,84 @@
...
+          key.nacl.verify(signature, data)
+          decoded_message
+        rescue RbNaCl::BadSignatureError
+          raise AuthenticationError.new('Token signature invalid')
+        end
+      end
+
+      private
+
+      def self.encode_message(message, footer)
```
Comment (mguymon) [inline lib/paseto/public.rb]: If this needs to be private, it has to be defined in 

    class << self
    end

but this is also a code smell resulting from relying on class methods to much.

File: lib/paseto/token.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+      message.join('.')
+    end
+  end
+
+  def self.parse_raw_token(raw, expected_header = nil, expected_footer = nil)
+    version, purpose, payload, footer = raw.split('.')
+
+    header = "#{version}.#{purpose}"
+    if !expected_header.nil? && header != expected_header
```
Comment (mguymon) [inline lib/paseto/token.rb]: In what scenario would the expected_header be `nil`, should it always been defined?

File: lib/paseto.rb [hunk 8]
```diff
@@ -43,13 +43,9 @@ def self.encode64(bin)
     Base64.strict_encode64(bin).gsub(/=+$/, '')
```
Comment (mguymon) [inline lib/paseto.rb]: (I did this) I just want to point out that I think this is a bit of a hack and would love verify actually works with other implementations. Not sure if there is a better way to do it and maintain compat with Ruby 2.3, newer Ruby has a flag to skip padding.

File: lib/paseto/token.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+        Paseto.encode64(payload)
+      ]
+
+      message << Paseto.encode64(footer) if footer
+
+      message.join('.')
+    end
+  end
+
+  def self.parse_raw_token(raw, expected_header = nil, expected_footer = nil)
```
Comment (mguymon) [inline lib/paseto/token.rb]: stylistic - class methods should be at the top of the class def

File: lib/paseto/public.rb [hunk 10]
```diff
@@ -1,54 +1,84 @@
...
-      computed_msg = Paseto.validate_and_remove_footer(message, @footer)
-      decoded_payload = Paseto.decode64(computed_msg[10..-1]);
+      class PublicKey
+        def self.decode64(encoded_key)
+          new(Paseto.decode64(encoded_key))
+        end
 
-      decoded_message = decoded_payload[0..-(SIGNATURE_BYTES + 1)]
-      signature = decoded_payload[-SIGNATURE_BYTES..-1]
+        def initialize(key)
```
Comment (mguymon) [inline lib/paseto/public.rb]: stylistic - should scoot this `initialize` below the class methods defs

File: spec/pasteo/local_spec.rb [hunk 11]
```diff
@@ -1,32 +1,70 @@
...
+    subject { described_class }
+
+    it '.encode64 returns a base64-encoded key' do
+      expect(key.encode64).to eq(encoded_key)
+    end
+
+    it '.decode64(key.encode64) is a no-op' do
+      key = subject.generate
+      round_trip = subject.decode64(key.encode64)
+      expect(round_trip.instance_variable_get(:@key)).to eq(key.instance_variable_get(:@key))
```
Comment (mguymon) [inline spec/pasteo/local_spec.rb]: So cracking open an instance for assertions is a code smell. Should be able to test its behavior without slipping around in the guts.

Maybe it makes sense to turn the `Key` classes in Structs and get the free equality check (assuming RbNaCl plays nice with `==`)?

File: spec/pasteo/public_spec.rb [hunk 12]
```diff
@@ -1,58 +1,104 @@
...
+      it "should raise when the footer doesn't match what's expected" do
+        expect do
+          subject.verify(signed_message_with_footer, public_key, bad_footer)
+        end.to raise_error Paseto::TokenError
       end
     end
   end
-end
\ No newline at end of file
+end
```
Comment (mguymon) [inline spec/pasteo/public_spec.rb]: Great additional tests :+1:

File: lib/paseto/token.rb (hunk 7 above)
Comment (AnIrishDuck) [reply lib/paseto/token.rb]: Missed this earlier, but `expected_header` is `nil` as part of `read_unauthenticated_footer`. I actually think I have a better way of doing this coming soon.

**Comment by AnIrishDuck** [comment]
@mguymon - There were quite a few changes needed to make the tokens fully "interoperable" (see 24134c1).

I also cleaned things up a bit more and responded to your feedback. The API is now centered around key and token objects, with "legacy" class methods for those who prefer that.

**Pull request merged**